File: src/java/4_BehavioralPatterns/5_Mediator/ChatRoom.java
package com.activemesa.behavioral.mediator;

import java.util.ArrayList;
import java.util.List;

class Person
{
  public String name;
  public ChatRoom room;
  private List<String> chatLog = new ArrayList<>();

  public Person(String name)
  {
    this.name = name;
  }

  public void receive(String sender, String message)
  {
    String s = sender + ": '" + message + "'";
    System.out.println("[" + name + "'s chat session] " + s);
    chatLog.add(s);
  }

  public void say(String message)
  {
    room.broadcast(name, message);
  }

  public void privateMessage(String who, String message)
  {
    room.message(name, who, message);
  }
}

class ChatRoom
{
  private List<Person> people = new ArrayList<>();

  public void broadcast(String source, String message)
  {
    for (Person person : people)
      if (!person.name.equals(source))
        person.receive(source, message);
  }

  public void join(Person p)
  {
    String joinMsg = p.name + " joins the chat";
    broadcast("room", joinMsg);

    p.room = this;
    people.add(p);
  }

  public void message(String source, String destination, String message)
  {
    people.stream()
      .filter(p -> p.name.equals(destination))
      .findFirst()
      .ifPresent(person -> person.receive(source, message));
  }
}

class ChatRoomDemo
{
  public static void main(String[] args)
  {
    ChatRoom room = new ChatRoom();

    Person john = new Person("John");
    Person jane = new Person("Jane");

    room.join(john); // no message here
    room.join(jane);

    john.say("hi room");
    jane.say("oh, hey john");

    Person simon = new Person("Simon");
    room.join(simon);
    simon.say("hi everyone!");

    jane.privateMessage("Simon", "glad you could join us!");
  }
}

File: src/swift/4_BehavioralPatterns/5_Mediator/ChatRoom.swift
import Foundation
import XCTest

class Person
{
  var name: String
  var room: ChatRoom? = nil // not in a room
  private var chatLog = [String]()

  init(_ name: String)
  {
    self.name = name
  }

  func receive(sender: String, message: String)
  {
    let s = "\(sender): `\(message)`"
    print("[\(name)'s chat session] \(s)")
    chatLog.append(s)
  }

  func say(_ message: String)
  {
    room?.broadcast(sender: name, message: message)
  }

  func pm(to target: String, message: String)
  {
    room?.message(sender: name, destination: target, message: message)
  }
}

class ChatRoom
{
  private var people = [Person]()

  func broadcast(sender: String, message: String)
  {
    for p in people
    {
      if p.name != sender
      {
        p.receive(sender: sender, message: message)
      }
    }
  }

  func join(_ p: Person)
  {
    let joinMsg = "\(p.name) joins the chat"
    broadcast(sender: "room", message: joinMsg)
    p.room = self
    people.append(p)
  }

  func message(sender: String, destination: String, message: String)
  {
    people.first{ $0.name == destination}?.receive(sender: sender, message: message)
  }
}

func main()
{
  let room = ChatRoom()

  let john = Person("John")
  let jane = Person("Jane")

  room.join(john)
  room.join(jane)

  john.say("hi room")
  jane.say("oh, hey john")

  let simon = Person("Simon")
  room.join(simon)
  simon.say("hi everyone!")

  jane.pm(to: "Simon", message: "glad you could join us!")
}

main()